### ISSUE TITLE
Remote Synchronization: What's the best way to handle entities with duplicate uuids?

### ISSUE THREAD

**Issue by cbaker6** [carekit-apple/CareKit#453]
When using CareKit, in particular the CareKitStore in it's current distributed way, I'm sure there will be minimal cases, if any of duplicate uuids being generated on an individual device. This is because the size of the uuids generated along with uniqueness only needing to be per entity. If the same uuid was generated for two different entities, it probably doesn't matter. I assume if a duplicate uuid was generated for the same entity type on the same device, CareKitStore has a way to mitigate or generate a new one without the user ever seeing the issue.

In the case of centralizing the data with the  `OCKRemoteSynchronizable`. It seems a clash of uuids are more likely to occur within a given entity/table, particularly when an application scales up to many users and devices. From the standpoint of a vender implementing `OCKRemoteSynchronizable`, ***it's simple enough to only accept an entity if it's uuid doesn't clash with the any other on the remote, but how should we handle rejecting/reporting a clash back to the CareKitStore?***

I've thought of some possible solutions, but they are not without limitation:
1. IMO, the best way I have thought of so far is to have a method that can be used through `OCKRemoteSynchronizable` to report "duplicate uuid" for a specific entity/uuid. CareKit would then treat the current uuid is "mutable" within the framework, and initiate a one time change to the uuid of that specific version of the entity (without triggering an update). From my understanding of CareKitStore, this seems like the easiest to do this as all entities that were related to the duplicate uuid are actually linked to the entity itself and not the uuid (the uuids are retrieved through getters/setters of the object itself), so it wouldn't trigger an update throughout the whole CareKitStore. After the uuid is changed, the CareKitStore attempts to synchronize the changed entity with OCKRemoteSynchronizable through it's standard methods. If the uuid change creates a version update or tombstone, it seems it will cause some of the issues in the following two solutions.
2. Another way, is that whenever a duplicate is detected on the remote, to trigger an update (or delete then add) of the entity with a duplicate uuid. This seems like it may cause a few issues, particularly for the versioned entities. This may work if the duplicate happens to be the newest version, but can be problematic if it's a previous version. This is because there will alway be a link to some entity that was can't be synced to the cloud. I guess it could be possible to assume the uuid on the remote isn't unique, but then you lose the ability of indexing the uuid on the remote here...
3. Another way is one I've kind of talked about before is leverage `remoteId` to hold the unique uuid given by the remote. This can possibly work, but means that after you first synchronize an entity to the remote, you trigger an update to a versioned entity (or tombstone to non-versioned) by adding the `remoteId` which then generates new uuids. So you really have two entities for every 1 created (I guess you could delete the first one). If the uuid generated from the update clashes with the remote, you are back to the issue of second scenario.

Of course, there are probably solutions that are not listed above. Do you have a suggested way to deal with this type of scenario? Do you think this a reasonable concern since the uuids are created in a distributed fashion?

**Comment by erik-apple** [comment]
Clashing UUIDs is technically possible since there are a finite number of UUIDs, but the probability is so mind bogglingly small that for practical purposes, you can typically assume collisions are impossible, even in large distributed systems.

There are lots of resources out there that provide the math if you're interested. [Wikipedia](https://en.wikipedia.org/w/index.php?title=Universally_unique_identifier&oldid=755882275#Random_UUID_probability_of_duplicates) is thorough, albeit not terribly reader friendly. There are lots of other articles, like [this one](https://towardsdatascience.com/are-uuids-really-unique-57eb80fc2a87), that are a bit more accessible. I like the explanation given at the end.

> A sample of 3.26*10¹⁶ UUIDs has a 99.99% chance of not having any duplicates. Generating that many UUIDs, at a rate of one per second, would take a billion years. So while UUIDs are not truly unique, they are unique enough for practical purposes, taking into account the natural limitations of human lifespans and separation of systems.

**Comment by cbaker6** [comment]
@erik-apple I see, I guess I always the assumed the low probability of clashing was due to uuids being generated in a centralized way, but it seems from the links that you sent, it should hold in the decentralized case. It also seems like the clashing is highly dependent on the random number generator of the OS, which I haven't heard of issues with in iOS

**Comment by erik-apple** [comment]
Yeah, if you assume that the implementation that creates the UUID's is correct, and that you don't need your UUIDs to be unique amongst all other systems (i.e. it's fine if a big box store somewhere tracks a shipment of shoes using a UUID you also use), then you should be fine. And I think both of those assumptions are fair to make!

**Comment by erik-apple** [comment]
Just to add a bit of clarification:

> I see, I guess I always the assumed the low probability of clashing was due to uuids being generated in a centralized way.

Your assumption was a good one. That is the case for many kinds of identifiers, particularly the `localVersionID`s that we recently moved away from. The `managedObjectID` generated by CoreData is guaranteed to be unique on a given device, but cannot be guaranteed to be unique across devices. This is primary reason we moved to using UUIDs instead, which are designed to be unique across devices.
